### PR TITLE
fix(es2018): add a target and an initial lib

### LINKED ISSUE #20342: Add an ES2018 Target and Lib
This should include any features we support which are slated for the 2018 spec release (currently stage 4 proposals), or things we think are likely to be in it (ie, promising stage 3 proposals).

### PR DESCRIPTION
**Pull request by benbraou**
With this PR:

1. jake LKG is run (first commit)
1. es2018 target and library is added (part of second commit)
1. jake LKG is run again (part of second commit) - 

Fixes #20342

### CODE REVIEW

File: src/compiler/commandLineParser.ts [hunk 1]
```diff
@@ -120,6 +121,7 @@ namespace ts {
                     "es7": "lib.es2016.d.ts",
                     "es2016": "lib.es2016.d.ts",
                     "es2017": "lib.es2017.d.ts",
+                    "es2018": "lib.es2018.d.ts",
```
Comment (mhegazy) [inline src/compiler/commandLineParser.ts]: you need to build the new library as well. this is defined in our Jake/gulp files (sorry we have two build systems for now untill we are done with the migration).
JakeFile: https://github.com/Microsoft/TypeScript/blob/8f1cdc9b0c649f1e4fe9d96718be93277baa2ecd/Jakefile.js#L229
GulpFle: https://github.com/Microsoft/TypeScript/blob/8f1cdc9b0c649f1e4fe9d96718be93277baa2ecd/Gulpfile.ts#L168
See https://github.com/Microsoft/TypeScript/commit/6a737c88f7734d7f6d70a85c11bf109bbe18436e that added esNext target.

Comment (benbraou) [reply src/compiler/commandLineParser.ts]: @mhegazy thank you for the review. I am updating my PR accordingly.

FYI, I see that jake LKG generates diagnosticMessages.generated.json file per locale. I think this is related to PR  #19308

Therefore, in my PR you will have 2 commits:
1. 'run jake LKG before es2018 addition' 
2. 'fix(es2018): add a target and an initial lib' (contains the es2018 lib + jake LKG run a second time)
Please let me know if I need to squash both commits

**Comment by mhegazy** [comment]
Thanks! For future references please do not include LKG updates in PRs, they happen as part of publishing or manually directly on master.

**Comment by mhegazy** [comment]
@benbraou do you want to pick up adding defintions for the new array methods: https://github.com/tc39/proposal-flatMap (a new `es2018.array.d.ts` file in src\lib) and Promise.prototype.finally https://github.com/tc39/proposal-promise-finally ( `es2018.promise.d.ts`)

**Comment by benbraou** [comment]
@mhegazy Thanks you! I got the point to not include LKG updates in future PRs.
Yes, I would love to add definitions for flatMap  and promise finally.
Do you want me to create 2 separate issues ? I will be having a look at the implementation tomorrow.

**Comment by mhegazy** [comment]
> Do you want me to create 2 separate issues ? 

that would be appreciated. thanks!

File: lib/lib.dom.iterable.d.ts [hunk 2]
```diff
@@ -107,6 +90,31 @@ interface NodeListOf<TNode extends Node> {
...
+     * Returns a list of keys in the list
+     */
+    keys(): IterableIterator<string>;
+    /**
+     * Returns a list of values in the list
+     */
+    values(): IterableIterator<string | File>;
+
+    [Symbol.iterator](): IterableIterator<string | File>;
+}
```
Comment (Kovensky) [inline lib/lib.dom.iterable.d.ts]: (out of scope for this PR but) this pattern repeats often enough that maybe there should be a virtual interface in `lib.es2015.iterable.d.ts` for this 🤔 

```ts
interface IterableCollection<K, V = K, I = V> {
  entries(): IterableIterator<[K, V]>;
  keys(): IterableIterator<K>;
  values(): IterableIterator<V>;
  [Symbol.Iterator](): IterableIterator<I>
}
```

File: lib/lib.es2015.core.d.ts [hunk 3]
```diff
@@ -69,10 +70,16 @@ interface ArrayConstructor {
...
+    from<T>(arrayLike: ArrayLike<T>): T[];
+
+    /**
+     * Creates an array from an iterable object.
+     * @param arrayLike An array-like object to convert to an array.
      * @param mapfn A mapping function to call on every element of the array.
      * @param thisArg Value of 'this' used to invoke the mapfn.
      */
-    from<T, U = T>(arrayLike: ArrayLike<T>, mapfn?: (v: T, k: number) => U, thisArg?: any): U[];
+    from<T, U>(arrayLike: ArrayLike<T>, mapfn: (v: T, k: number) => U, thisArg?: any): U[];
```
Comment (Kovensky) [inline lib/lib.es2015.core.d.ts]: `from<T, U, This>(arrayLike: ArrayLike<T>, mapfn: (this: This, v: T, k: number) => U, thisArg?: This): U[]`?

**Pull request merged**